### PR TITLE
Performance Tweaks

### PR DESCRIPTION
**Pull request by Slumpwuffle**
Adds a number of performance tweaks to the loading process. This brings down the load for fishtornado on a Quest2 headset from ~30-40s to ~5-6s.

### CODE REVIEW

File: include/ppx/bitmap.h
Comment (apazylbe) [inline include/ppx/bitmap.h:201]: I think the project convention is to name these pWidth, etc

**Review by apazylbe — COMMENTED** [review]
Thanks for the PR! Would you be able to pull in latest changes? There are some conflicts now

File: src/ppx/bitmap.cpp
Comment (Slumpwuffle) [reply src/ppx/bitmap.cpp]: Switched this back to a reference. Style preference turned out to be 'pointer only if the argument is optional', which isn't the case here.

File: src/ppx/fs.cpp
Comment (Slumpwuffle) [reply src/ppx/fs.cpp:119]: ah ew, cruft I forgot to clean up while experimenting with another approach. Done.

File: include/ppx/mipmap.h
Comment (Slumpwuffle) [reply include/ppx/mipmap.h:44]: Done.

File: include/ppx/bitmap.h
Comment (Slumpwuffle) [reply include/ppx/bitmap.h]: Done.

File: src/ppx/bitmap.cpp
Comment (chaoticbob) [inline src/ppx/bitmap.cpp:588]: nit: pointer should have prefixed `p`.

**Review by chaoticbob — COMMENTED** [review]
Looks good overall. Just some nits.

File: src/ppx/bitmap.cpp
Comment (Slumpwuffle) [reply src/ppx/bitmap.cpp]: I think this was residue from an earlier experiment; removed and doesn't appear to affect compilation.

File: src/ppx/tri_mesh.cpp
Comment (Slumpwuffle) [reply src/ppx/tri_mesh.cpp]: Done. Note there are some non-compliant functions around, e.g. ppx::fs::path_exists.

File: src/ppx/bitmap.cpp
Comment (Slumpwuffle) [reply src/ppx/bitmap.cpp]: Done.

**Review by chaoticbob — COMMENTED** [review]
LGTM but lets wait to see if @Keenuts has any feedback before merging.

**Review by Keenuts — COMMENTED** [review]
Sorry for the delay!
I see 3 optimizations:
 1 - smarter memory usage for temporary images
 2 - avoiding copying stbi buffers when possible
 3 - replacing jumps with indirect calls

1. Yes, makes perfect sense, as the pattern is always map, load, upload-GPU, unmap. LGTM on this (just some nits)

2. Yes, makes sense. We might even be able to go further if we use spans no?

3. Not sure. Is this change alone actually causing a notable perf increase? Why is that better?
In the first case, you have a cmp+jmp in the loop. But in the new code you have a load+call.
Because those are indirect calls, I'm guessing the compiler cannot go full-naked function, and still has to respect some calling conventions, hence may move stuff to specific registers, save the stack, etc. While the previous code didn't have any of that.

File: src/ppx/mipmap.cpp
Comment (chaoticbob) [reply src/ppx/mipmap.cpp:92]: Probably not guranteed, it's just an allocation underneath. We haven't done anything big enough to hit any memory limits. Should probably have an additional assert tho, the return isn't really meaningful since we don't catch exceptions.

Comment (Slumpwuffle) [reply src/ppx/mipmap.cpp:92]: yeah, I was just duplicating the existing logic, but I don't think it's correct. Looking online, there isn't defined behavior for out of memory issues, but generally a bad_alloc exception should be thrown. Some other implementations simply overrun memory. But there's no definition that indicates a smaller size than requested would be returned.

I think this is one of those 'it's safe to just assume it gives you the memory you ask for, because otherwise it'll just crash' situations. Will remove this logic.

File: src/ppx/mipmap.cpp
Comment (Slumpwuffle) [reply src/ppx/mipmap.cpp:70]: You mean within the class? Static members cannot be initialized within the header since it would initialize across all TUs including the header. They need to be initialized within a specific TU, generally the CPP file for the header.

**Comment by Slumpwuffle** [comment]
Okay, github doesn't let me reply to your comment directly, but:
>Not sure. Is this change alone actually causing a notable perf increase? Why is that better?
In the first case, you have a cmp+jmp in the loop. But in the new code you have a load+call.
Because those are indirect calls, I'm guessing the compiler cannot go full-naked function, and still has to respect some calling conventions, hence may move stuff to specific registers, save the stack, etc. While the previous code didn't have any of that.

The answer is that branches are indeed slower. Admittedly most of the gains are from memory allocation, but branches require pipeline and cache flushes if mispredicted, and often pipeline stalls even if not. Static flows are almost always faster even with the additional memory access and instructions because those will end up in cache. On what was a 7s load, the branch removal was roughly a 1s gain, the memory allocs were the rest of the 6s.

File: src/ppx/mipmap.cpp
Comment (Keenuts) [reply src/ppx/mipmap.cpp:70]: Oh sorry, my comment was on the wrong line, I means for the overloaded function:
```
Mipmap::Mipmap(uint32_t width, uint32_t height, Bitmap::Format format, uint32_t levelCount)
Mipmap::Mipmap(uint32_t width, uint32_t height, Bitmap::Format format, uint32_t levelCount, bool useStaticPool)
```

File: src/ppx/fs.cpp
Comment (Keenuts) [reply src/ppx/fs.cpp:118]: That's a fair point, ok with that.

File: src/ppx/mipmap.cpp
Comment (Keenuts) [reply src/ppx/mipmap.cpp]: Yeah, I cannot find exact rule saying if it should always except or not.  It says there are no effect if an exception is thrown, but not if the allocation can silently fails. Seems like implems I've seen do throw with default allocators, so would agree removing the check should be enough.

**Comment by Keenuts** [comment]
> The answer is that branches are indeed slower. Admittedly most of the gains are from memory allocation, but branches require pipeline flushes if mispredicted or lacking prediction, and often pipeline stalls even if not. Static flows are almost always faster even with the additional memory access and instructions because those will end up in cache. On what was a 7s load, the branch removal was roughly a 1s gain, the memory allocs were the rest of the 6s.

So you have measurements to back this? Branch prediction can indeed be wrong, so is speculative execution. (Since we are not at a ring boundary, I believe speculative execution is still authorized and unhindered by all the spectre mitigations and such).
Anyway, I'm definitely not an expert of micro-architectural details as they are way too complex nowadays.

What I did on my end was to keep all the PR changes (including the pre-allocation from "Speed up obj-based mesh loading"), added some timers and did multiple headless runs through perf  with 0 frames on 15_basic_material. What I compared was the PR as it is, and PR without the branch fix. Optimization enabled of course, as comparing without wouldn't make much sense.
It's on my workstation, so obviously some noise, so added the margins I observed.

```
With the branch fix:
page_faults  : average=54259         (+- 20662)
icache_miss  : average=1.2M          (+- 184K)
cache_miss   : average=199M          (+- 26M)
branch_miss  : average=65M           (+- 1M)
branch       : average=4B            (+- 1M)
instruction  : average=26B           (+- 4M)
model        : average=0.896s        (+- 0.020s)
texture      : average=2.742s        (+- 0.106s)

Without the branch fix:
page_faults  : average=55768         (+- 10337)
icache_miss  : average=636K          (+- 22K)
cache_miss   : average=16M           (+- 3M)
branch_miss  : average=65M           (+- 815K)
branch       : average=4B            (+- 527K)
instruction  : average=22B           (+- 3M)
model        : average=0.102         (+- 0.007s)
texture      : average=2.739         (+- 0.070s)
```

(Raw results in case you want to have more precision)
```
With branch fix (pr as is):
page_faults  : average=54259.083	(margin= 20662.000)
icache_miss  : average=1180922.500	(margin=184062.000)
cache_miss   : average=199257846.667	(margin=26665308.000)
branch_miss  : average=65546041.750	(margin=976332.000)
branch       : average=4596896371.667	(margin=907387.000)
instruction  : average=26363841984.500	(margin=4253171.000)
model        : average=0.896	(margin=     0.020)
texture      : average=2.742	(margin=     0.106)

Without branch fix (only smarter allocations):
page_faults  : average=55768.833	(margin= 10337.000)
icache_miss  : average=636745.833	(margin= 22821.000)
cache_miss   : average=16867040.500	(margin=3138835.000)
branch_miss  : average=65119469.583	(margin=815188.000)
branch       : average=4234112410.167	(margin=527591.000)
instruction  : average=22350433883.250	(margin=3907268.000)
model        : average=0.102	(margin=     0.007)
texture      : average=2.739	(margin=     0.070)
```

**Comment by chaoticbob** [comment]
If there is a noticable performance improvement for Android, I'd like to get some of these changes in soon and worry about the specifics of the benchmarking later (assuming they're not egregious). vk_15_basic_materials takes minutes to load currently on Android and I'd like for the Android fixes to not conflict with this PR.

**Comment by Slumpwuffle** [comment]
Yep, okay, I agree, it's not the branching. I only saw the difference in one run before, and it could've been inconsistent alloc behavior. Trying it now isolating the branching vs. functional variants, there's no meaningful net impact (seeing ~0.048-0.049 seconds for loading the shark each time regardless of approach). I think maybe losing the branches is faster on average by about ~0.001s, but it's likely not worth the code churn so I'll revert it.

File: include/ppx/tri_mesh.h
Comment (Keenuts) [inline include/ppx/tri_mesh.h:165]: comment maybe? Bigwheels hasn't been great on comments, but we are trying to change that, at least in headers.

**Review by Keenuts — APPROVED** [review]
Some comments left to fix, mostly about comments.
Otherwise fine with the change.
For comments we haven't been that great in BigWheels so far, but let's try to at least comment prototypes in headers.

File: include/ppx/mipmap.h
Comment (Slumpwuffle) [reply include/ppx/mipmap.h]: I feel that the term "static pool" ought to be clear enough on its own under most circumstances, but recognize folks might not understand the definition. Done.

File: include/ppx/bitmap.h
Comment (Slumpwuffle) [reply include/ppx/bitmap.h]: Clarified format, mentioned the rest mirror stbi_load.

File: include/ppx/tri_mesh.h
Comment (Slumpwuffle) [reply include/ppx/tri_mesh.h:167]: Added comment.

File: src/ppx/bitmap.cpp
Comment (Slumpwuffle) [reply src/ppx/bitmap.cpp]: Cleaned up both sites.

**Pull request merged**
